### DISCUSSION TITLE
Training data explanation

### DISCUSSION THREAD

**Discussion by zeyunZhao** [wmglab-duke/axonml#5] [General]
Hi, this is a nice job! 

I have been trying to reproduce your work on GPU again, but I find it is very hard to understand the meanings of your model's inputs and outputs. (probably bc I am a freshman in this domain). Could you please tell me what format of training data are supposed to be fed into the axonml model? I just know their size but it is really hard to know their meanings of their names and what each channel represents. Could you give me more details for me to understand your model better? Thank you so much!

activations :   (1, 1, 32)

m :   (1, 1, 32, 53, 1001)
h :   (1, 1, 32, 53, 1001)
p :   (1, 1, 32, 53, 1001)
s :   (1, 1, 32, 53, 1001)
v :   (1, 1, 32, 53, 1001)

diameters :   (1, 32, 32)
parameters :   (1, 1, 64)
e :   (1, 1, 32, 53, 1001)

**Comment by zeyunZhao** [comment]
you mentioned there are 26 parameters of S-MF were optimized, but when I counted it, there are only 23 parameters: 1> maximumionic  conductancevalues(gNaf,gNap,gK,andgL)--4 parameters;  2>  axialintracellular resistivity (ρa) --1 parameter; 3> membranecapacitance(cm)--1 parameter; 4> coefficients of the dnode interpolant (βn 1,βn 2,βn 3)--3 parameters; 5>  coefficientsof thedaxon interpolant (βa 1,βa 2,βa 3)--3 parameters; 6> q10 parameters (aqm 10,aqh 10,aqp 10,aqs 10)--4 parameters; 7> governing K+ dynamics--3 parameters; 8> thevaluesof the symmetric convolutional kernel ofCNN(4values)--4 parameters; In sum: 4+1+1+3+3+4+3+4=23. 

And when I read codes, the size of parameters are 64. It is really confusing.

**Comment by zeyunZhao** [comment]
I have asked GPT for the illustration of these parameters and get answers as bellows. Could you tell me if they are right and could you tell me what is the parameters with size of [1,1,64]? 

---------------------------------------------------------------
In the context of the McIntyre-Richardson-Grill (MRG) model of neural axons, each parameter represents different states or components of axonal membranes and ion channels across different nodes of Ranvier and myelinated segments. Let’s go through each parameter and what it represents:

activations (1, 1, 32)

Represents activation states or levels across 32 different axonal nodes. This typically might be a simplified representation of action potential generation or readiness in each node.
m, h, p, s, v (1, 1, 32, 53, 1001)
These represent variables associated with different ion channel states across the nodes and internodal regions of the axon. Here, (1, 1, 32, 53, 1001) suggests the following:

32: Total axonal segments/nodes being simulated.

53: Discretization of spatial points along the length of each node or internodal region.

1001: Number of time points across the simulation.

m: Represents the activation gate of sodium channels. This variable controls how many sodium channels are open, affecting the inflow of Na+  ions, which is crucial for action potential initiation.

h: Represents the inactivation gate of sodium channels. The inactivation gate closes the sodium channels after depolarization, preventing Na+  ions from entering the cell during the refractory period.

p: Represents the activation gate of potassium channels. This variable manages the K+  ion outflow, which helps repolarize the membrane after an action potential.

s: Represents the activation of slow potassium channels or other delayed rectifier currents. This helps in fine-tuning the repolarization phase and can contribute to action potential duration.

v: Represents the membrane potential across different spatial locations and time points. This value changes dynamically during action potentials and is central to the axon's excitability.

diameters (1, 32, 32)

This likely represents the diameter of each segment of the axon, which affects the speed and efficiency of action potential propagation. The three dimensions (1, 32, 32) suggest the diameters could vary along the 32 nodes or segments of the axon.
parameters (1, 1, 64)

This parameter contains model-specific constants and parameters such as conductance values, capacitances, ion concentrations, or other biophysical properties that influence axonal behavior. These may be used to initialize or control the simulations.
e (1, 1, 32, 53, 1001)

Likely represents extracellular potential or field effects influencing the axonal membrane at different spatial points and time steps. It can account for extracellular stimuli or background electric fields affecting axonal excitability.
Each of these variables is essential in simulating the electrical activity of a myelinated axon realistically in the MRG model, reflecting various ionic currents, gating mechanisms, and membrane potential changes during action potentials across nodes of Ranvier and internodes.
----------------------------------------------------------------
Any reply from you will be appreciated!

**Comment by minhajh** [comment]
I moved this to Discussion as it's not addressing a (potential) bug / issue with the implementation.

The ChatGPT summary is generally OK. m, h and p are all states associated with Na+ conductances (m and h are fast and p is persistent). Only s is related to K+. Check the MRG paper (https://doi.org/10.1152/jn.00353.2001).

"activations" is not used in training - it records whether the axon in the dataset generated an action potential or not.

"parameters" is also not used in training - it records the parameter values that were run to generate the dataset.

There are 26 optimized parameters - there are 6 parameters associated with K+ dynamics (the alpha and beta equations governing s), not 3.

Hope this helps!
